### ISSUE TITLE
Programmatically solution for naming of "media-type-supported" tags

### ISSUE THREAD

**Issue by JSEHV** [attah/harbour-seaprint#60]
At the moment the "media-type-supported” tags are not nicely showing up in SeaPrint in a human friendly format and I (and probably you too) would prefer to see all Supported Media Types in a human friendly way.
I’ve analysed all the supplied IPP logs in the SeaPrint thread on the Sailfish forum and also all my in the past supplied IPP logs of printers regarding the "media-type-supported” tags.

What I found out analysing all these tags in the IPP logs is that there is a way we can solve this programmatically. 
To get there, we need to (I think in this exact order):

- remove the first “com.” (Including the dot) when found at the start of the string (https://thispointer.com/how-to-remove-substrings-from-a-string-in-c/)
- replace all the dashes “-“ and dots “.” with a space “ " (https://www.cplusplus.com/reference/string/string/replace/)
- add a space “ “ before each capital and number found, starting from the second position and when the next character is not a capital or a number (to get HP Matte 90gsm… and HP Eco SMART Lite…)
- capitalise the first letter of each word (https://www.w3resource.com/cpp-exercises/string/cpp-string-exercise-3.php)

(Optional)
- replace “Hp" with “HP” and “Oki” with “OKI” to be exact on the brands (I don’t know if people would care to this level, I even think Hp and Oki looks well :) )

We should end up with a very nice list in a human friendly format. See the [Media Type Supported.pdf](https://github.com/attah/harbour-seaprint/files/7861946/Media.Type.Supported.pdf) file for the result, if I am correct.

As you can see in the PDF-file, also the first 3 photographic types will change (current is in between brackets), but still very understandable and in my view perfectly correct.
The nice thing if this all is, whatever the amount of strings is thrown to SeaPrint via IPP tags by printers, SeaPrint will all solve it nicely in a programmatical way (jippie!) :)

Hope you like the solution and are able to implement it.

**Comment by attah** [comment]
Interesting idea. This is certainly in the "good enough" spirit of SeaPrint.
However; there is a "proper" way to do it that i have not explored yet - both due to not having a printer supporting it, and priorities.
It would also address the obvious issue here - not being able to translate the strings (not that it is enabled currently).

The printer actually provides a dictionary for translation - and more relevant in this case, explanation.
See "printer-strings-uri". Could you please send me your printer's dictionary so i can try?
That doesn't mean it must work this way, but i should probably at least try.

Another thing i have considered is to collect the 100 or so most common private media types... should have a similar, but less fun effect.
Anyway; currently the lookup table is here: https://github.com/attah/harbour-seaprint/blob/6095694d92ccafd076f13a56b85136e962d0e943/qml/pages/utils.js#L733
Probably programmatic prettification would also stay in javascript, but getting the printer dictionary would certainly have at least some c++.
The reason for some seemingly common names missing is that they appear to not be in the spec https://www.iana.org/assignments/ipp-registrations/ipp-registrations.xhtml

**Comment by JSEHV** [comment]
Printer supports more languages, but I supplied the ones where [SeaPrint currently has translations](https://github.com/attah/harbour-seaprint/tree/master/translations) available

English
```
"media.om_hivision_101.6x180.6mm" = "16:9 wide (4 x 7.11 in)";
"media.om_4psize_254x305mm" = "25.4 x 30.5 cm (10 x 12 in)";
"media.oe_hansetsu_14x17in" = "35.6 x 43.2 cm (14 x 17 in)";
"media-type.com.epson-photographic-glossy" = "Epson Premium Glossy";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Washi";
"media-type.com.epson-matte-fine-art" = "Fine Art Matte";
"media-type.com.epson-matte-canvas" = "Canvas Matte";
"media-type.com.epson-satin-canvas" = "Canvas Satin";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";

```
German
```
"media.om_hivision_101.6x180.6mm" = "16:9 breit (4 x 7,11 Zoll)";
"media.om_4psize_254x305mm" = "25,4 x 30,5 cm (10 x 12 Zoll)";
"media.oe_hansetsu_14x17in" = "35,6 x 43,2 cm (14 x 17 Zoll)";
"media-type.com.epson-photographic-glossy" = "Epson Premium Glossy";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Washi";
"media-type.com.epson-matte-fine-art" = "Fine Art Matte";
"media-type.com.epson-matte-canvas" = "Leinwand Matte";
"media-type.com.epson-satin-canvas" = "Leinwand Satin";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";
```
French
```
"media.om_hivision_101.6x180.6mm" = "16:9 large (4 x 7,11 p.)";
"media.om_4psize_254x305mm" = "25,4 x 30,5 cm (10 x 12 p.)";
"media.oe_hansetsu_14x17in" = "35,6 x 43,2 cm (14 x 17 p.)";
"media-type.com.epson-photographic-glossy" = "Epson Premium GlacÃ©";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Papier Japonais";
"media-type.com.epson-matte-fine-art" = "Papier beaux-arts mat";
"media-type.com.epson-matte-canvas" = "Toile Canvas mate";
"media-type.com.epson-satin-canvas" = "Toile Canvas satinÃ©e";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";
```
Spanish
```
"media.om_hivision_101.6x180.6mm" = "16:9 ancho (4 x 7,11 pulg)";
"media.om_4psize_254x305mm" = "25,4 x 30,5 cm (10 x 12 pulg)";
"media.oe_hansetsu_14x17in" = "35,6 x 43,2 cm (14 x 17 pulg)";
"media-type.com.epson-photographic-glossy" = "Epson Premium Glossy";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Papel Washi";
"media-type.com.epson-matte-fine-art" = "Papel para Bellas Artes";
"media-type.com.epson-matte-canvas" = "Lienzo mate";
"media-type.com.epson-satin-canvas" = "Lienzo satinado";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";
```
Polish
```
"media.om_hivision_101.6x180.6mm" = "Panoramiczny 16:9 (4 x 7,11 cala)";
"media.om_4psize_254x305mm" = "25,4 x 30,5 cm (10 x 12 cali)";
"media.oe_hansetsu_14x17in" = "35,6 x 43,2 cm (14 x 17 cali)";
"media-type.com.epson-photographic-glossy" = "Epson Premium Glossy";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Washi";
"media-type.com.epson-matte-fine-art" = "Fine Art Matte";
"media-type.com.epson-matte-canvas" = "Canvas Matte";
"media-type.com.epson-satin-canvas" = "Canvas Satin";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";
```
Dutch
```
"media.om_hivision_101.6x180.6mm" = "16:9 breedbeeldformaat (4 x 7,11 in)";
"media.om_4psize_254x305mm" = "25,4 x 30,5 cm (10 x 12 in)";
"media.oe_hansetsu_14x17in" = "35,6 x 43,2 cm (14 x 17 in)";
"media-type.com.epson-photographic-glossy" = "Epson Premium Glossy";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Washi";
"media-type.com.epson-matte-fine-art" = "Fine Art Matte";
"media-type.com.epson-matte-canvas" = "Canvas Matte";
"media-type.com.epson-satin-canvas" = "Canvas Satin";
"media-type.com.epson-coated" = "Epson Photo Quality Ink Jet";
```
Chinese (zh_cn)
```
"media.om_hivision_101.6x180.6mm" = "16:9 å®½å± (4 x 7.11 è‹±å¯¸)";
"media.om_4psize_254x305mm" = "254 x 305 æ¯«ç±³ (10 x 12 è‹±å¯¸)";
"media.oe_hansetsu_14x17in" = "356 x 432 æ¯«ç±³ (14 x 17 è‹±å¯¸)";
"media-type.com.epson-photographic-glossy" = "çˆ±æ™®ç”Ÿé«˜è´¨é‡å…‰æ³½çº¸";
"media-type.com.epson-luster" = "Epson Premium Luster Photo Paper";
"media-type.baryta" = "Baryta";
"media-type.com.epson-proofing-white-semimatte" = "Epson Proofing Paper White Semimatte";
"media-type.com.epson-velvet-fine-art" = "Velvet Fine Art Paper";
"media-type.com.epson-fine-art" = "Watercolor Paper Radiant White";
"media-type.com.epson-ultra-smooth-fine-art" = "Epson UltraSmooth Fine Art paper ";
"media-type.washi" = "Washi";
"media-type.com.epson-matte-fine-art" = "Fine Art Matte";
"media-type.com.epson-matte-canvas" = "Canvas Matte";
"media-type.com.epson-satin-canvas" = "Canvas Satin";
"media-type.com.epson-coated" = "çˆ±æ™®ç”Ÿç…§ç‰‡è´¨é‡å–·å¢¨çº¸";
```

**Comment by attah** [comment]
Thanks. I'll just use the one as an example, and let SeaPrint follow what the printer is configured to.
They should all have the same format...

**Comment by JSEHV** [comment]
I am looking forward to it :) Hope it can make v1.0.4

**Comment by attah** [comment]
I almost forgot.. 
After sleeping on it, strings support wasn't so much work.
Please test https://github.com/attah/harbour-seaprint/releases/tag/1.0.4-beta1

**Comment by JSEHV** [comment]
Working well as far as I can see (see attached screenshots). So if you release 1.0.4, you can close this thread.

The only question I would have is, is the ink levels in the middle looking balanced on the printer details page? To me, first all the text and still on the bottom the supply levels will look better (how it was before). I didn't expect this, but in the looks of design it breaks too much in my opinion. Maybe when you make tabs for it (like you do with media sections when printing something) is an option. If you do that for both the text part of the details as for the supply levels (and both have them opened by default), you can with one click close the text part insted of scrolling down to the supply levels. 

But it is up to you if you think that will work for you and what you intend to do.

Besides that, another great release!

![Screenshot_20220117_001](https://user-images.githubusercontent.com/5128988/149776466-25ebf37f-4e6d-4541-b150-d98e17417f98.png)
![Screenshot_20220117_002](https://user-images.githubusercontent.com/5128988/149776474-d1e07823-4d91-4a68-a84c-2a0c1888d95b.png)
![Screenshot_20220117_003](https://user-images.githubusercontent.com/5128988/149776485-90ef4be6-dee6-497d-b2a1-7d651b0e36e2.png)

**Comment by attah** [comment]
Good, happy to hear.
I'll try to not hurry too much with the next release, there is always something right the day after if i do.
(But shouldn't be too long)

For printers with 4-5 supplies i think it is just fine this way, but for your printer i see what you mean.
Perhaps after status, supplies are probably most important so i want them placed prominently, both for usage and screenshots (showing off features).
I don't think collapsible sections are the solution here - tabs like in settings/phone/clock might be worth exploring.
Also i think scrolling/swiping is faster/easier than collapsible sections.
IMO details are supposed to be detailed, so i think another details button would be a bit silly.

I'll put it in the back of my mind and see what shakes out.

**Comment by JSEHV** [comment]
Tabs are a great idea to do! Keeps it even more clean in my view.
Plus it wil fit even more to the Sailfish OS design style. 

Would that be an option too for the 2 Media sections currently on the bottom as collapsible sections when printing?

**Comment by attah** [comment]
Let's see how that goes. :)

It is not an option for the job options, it would break going back to the printer list.
There i'm quite happy with hiding the advanced options as they would be less supported, and less used.

**Comment by attah** [comment]
Hmm, seems it is private :(
I guess i have some more thinking to do.

**Comment by JSEHV** [comment]
Got feedback from another dev (Nico):

<Nico> I think the tab API is still private. But you can probably just look at the qml files for it and copy it.
<Me> With private it means only Jolla can use it for now?
<Nico> You can use it too, but it is undocumented and not allowed to be published in harbour and it might break at any moment.

So for now maybe not yet a solution. Hope later when it becomes a public api.

**Comment by attah** [comment]
I just shipped 1.0.4 as is. 
I do want to make the details page prettier, but it is at least quite functional... so it can wait until i figure out how.
